tool/tsh/common/tsh.go

```diff
@@ -4042,13 +4055,49 @@ func onSSH(cf *CLIConf) error {
 	}
 
 	tc.Stdin = os.Stdin
+
+	// Handle fork after authentication.
+	if cf.ForkAfterAuthentication && cf.forkSignalFd == 0 {
+		if len(cf.RemoteCommand) == 0 {
+			return trace.BadParameter("fork after authentication not allowed for interactive sessions")
+		}
+		forkParams := client.ForkAuthenticateParams{
+			GetArgs: func(signalFd uint64) []string {
+				return append([]string{
+					// --fork-signal-fd goes immediately after `tsh`.
+					"--fork-signal-fd", strconv.FormatUint(signalFd, 10),
+				}, cf.rawArgs...)
```

Comment:
We are doing a whooooole lot of things before ever reaching the point where we check if we are in a reexecuted background `tsh ssh`, which makes me quite uncomfortable. Can we have this be a dedicated command instead, in some way? Something that hooks in `main()` much earlier?

Reply:
Also, won't we just rerun the same host ID resolution twice? What if the results change and now we have two hosts when we used to have one?

If we have resolved a label query or a hostname passed through proxy templates into a host ID we should not redo the resolution again (and we should not connect to the auth if not necessary).

Reply:
I moved the reexec to before we make any clients. I'm hesitant to have a dedicated command; that's a whole set of flags that has to stay in sync with `tsh ssh`, and rewriting the original args to the new command would require using the parsed args in kingpin, which is much more complicated than just sticking the new flags at the beginning.